modules/trustxBidAdapter.js

```diff
@@ -0,0 +1,210 @@
+    noBid: 'Array of bid objects is empty',
+    noPlacementCode: 'Can\'t find placementCode for bid with auid - ',
+    havePCodeFor: ', placementCode is available only for the following uids - ',
+    emptyUids: 'Uids should be not empty',
+    emptySeatbid: 'Seatbid array from response has empty item',
+    emptyResponse: 'Response is empty',
+    hasEmptySeatbidArray: 'Response has empty seatbid array',
+    hasNoArrayOfBids: 'Seatbid from response has no array of bid objects - '
+  };
+
+  function _makeHandler(auids, placementMap, onTimeoutWorker, needToSendStat) {
+    var cbName = bidderCode + '_callback_wrapper_' + auids.join('_');
+    $$PREBID_GLOBAL$$[cbName] = function(resp) {
+      delete $$PREBID_GLOBAL$$[cbName];
+      _responseProcessing(resp, auids, placementMap, onTimeoutWorker, needToSendStat);
+    };
+    return '$$PREBID_GLOBAL$$.' + cbName;
+  }
+
+  function _sendRequest(auids, placementMap, onTimeoutWorker) {
+    var query = [];
+    var path = reqPath;
+    var needToSendStat = Math.random() < 0.1;
+    query.push('u=' + encodeURIComponent(location.href));
+    query.push('auids=' + encodeURIComponent(auids.join(',')));
+    query.push('cb=' + _makeHandler(auids, placementMap, onTimeoutWorker, needToSendStat));
+    query.push('pt=' + (window.globalPrebidTrustxPriceType === 'net' ? 'net' : 'gross'));
+    query.push('hbstat=' + (needToSendStat ? '1' : '0'));
+
+    adloader.loadScript(reqHost + path + query.join('&'));
+  }
+
+  function _makeTimeoutWorker() {
+    var active = true;
+    var handler = function() {
+      worker.called = true;
+      $$PREBID_GLOBAL$$.offEvent('bidTimeout', handler);
+    };
+    $$PREBID_GLOBAL$$.onEvent('bidTimeout', handler);
+    var worker = {
+      startTime: new Date().getTime(),
+      off: function() {
+        if (active && !this.called) {
+          active = false;
+          $$PREBID_GLOBAL$$.offEvent('bidTimeout', handler);
+        }
+      },
+      called: false
+    };
+    return worker;
+  }
+
+  function _callBids(params) {
+    var auids = [];
+    var placementMap = {};
+    var hasBid;
+    var bid;
+    var bids = params.bids || [];
+    for (var i = 0; i < bids.length; i++) {
+      bid = bids[i];
+      if (bid && bid.bidder === bidderCode && bid.placementCode) {
+        hasBid = true;
+        if (bid.params && bid.params.uid) {
+          if (!placementMap[bid.params.uid]) {
+            placementMap[bid.params.uid] = [bid.placementCode];
+            auids.push(bid.params.uid);
+          } else {
+            placementMap[bid.params.uid].push(bid.placementCode);
+          }
+        }
+      }
+    }
+
+    if (auids.length) {
+      var onTimeoutWorker = _makeTimeoutWorker();
+      _sendRequest(auids, placementMap, onTimeoutWorker);
+    } else if (hasBid) {
+      utils.logError(LOG_ERROR_MESS.emptyUids);
+    }
+  }
+
+  function tryToSendStat(startTime, timeout, auids, isTimeouted, error) {
```

Comment:
We don't permit bid adapters to collect information about the auction mechanics, this function should be dropped. If you'd like to collect this info please create an analytics adapter

Reply:
This makes sense. I would like to note that this function doesn't collect any information on the 'auction mechanics' itself- it sends TrustX header bidder information on the amount of header bids which have timeouted. Do you still think we should remove this function from the adapter code? Would an analytics adapter provide such information to the header bidder?

Reply:
This makes sense. I would like to note that this function doesn't collect any information on the 'auction mechanics' itself- it sends TrustX header bidder information on the amount of header bids which have timeouted. Do you still think we should remove this function from the adapter code? Would an analytics adapter provide such information to the header bidder?

Reply:
Thanks for clarifying. Yes, the function should remain removed as adapters should only make the minimum amount of network calls necessary for the sake of performance. Analytic adapters do have access to the `BID_TIMEOUT` event

Reply:
Thanks for the details. As I can see you have removed the 'trytosend' stat function. Do I understand correct that you don't need us to make any changes to the TrustX adapter code for now?

Reply:
Correct, no further changes requested from me. Next a second reviewer will be assigned per our PR process and they may have additional feedback for you before merging